File: clients/ocaml/lib/protocol.ml
(** Minimal msgpack encoder/decoder for kgd msgpack-RPC. *)

type t =
  | Nil
  | Bool of bool
  | Int of int
  | Str of string
  | Bin of bytes
  | Array of t list
  | Map of (t * t) list

(* -------------------------------------------------------------------------- *)
(* Encoding                                                                   *)
(* -------------------------------------------------------------------------- *)

let buf_add_uint8 buf v = Buffer.add_char buf (Char.chr (v land 0xff))

let buf_add_uint16_be buf v =
  Buffer.add_char buf (Char.chr ((v lsr 8) land 0xff));
  Buffer.add_char buf (Char.chr (v land 0xff))

let buf_add_uint32_be buf v =
  Buffer.add_char buf (Char.chr ((v lsr 24) land 0xff));
  Buffer.add_char buf (Char.chr ((v lsr 16) land 0xff));
  Buffer.add_char buf (Char.chr ((v lsr 8) land 0xff));
  Buffer.add_char buf (Char.chr (v land 0xff))

let encode_int buf n =
  if n >= 0 then begin
    if n <= 0x7f then
      (* positive fixint *)
      buf_add_uint8 buf n
    else if n <= 0xff then begin
      (* uint 8 *)
      buf_add_uint8 buf 0xcc;
      buf_add_uint8 buf n
    end
    else if n <= 0xffff then begin
      (* uint 16 *)
      buf_add_uint8 buf 0xcd;
      buf_add_uint16_be buf n
    end
    else begin
      (* uint 32 *)
      buf_add_uint8 buf 0xce;
      buf_add_uint32_be buf n
    end
  end
  else begin
    if n >= -32 then
      (* negative fixint *)
      buf_add_uint8 buf (n land 0xff)
    else if n >= -128 then begin
      (* int 8 *)
      buf_add_uint8 buf 0xd0;
      buf_add_uint8 buf (n land 0xff)
    end
    else if n >= -32768 then begin
      (* int 16 *)
      buf_add_uint8 buf 0xd1;
      buf_add_uint16_be buf (n land 0xffff)
    end
    else begin
      (* int 32 *)
      buf_add_uint8 buf 0xd2;
      buf_add_uint32_be buf (n land 0xffffffff)
    end
  end

let encode_str buf s =
  let len = String.length s in
  if len <= 31 then begin
    (* fixstr *)
    buf_add_uint8 buf (0xa0 lor len)
  end
  else if len <= 0xff then begin
    (* str 8 *)
    buf_add_uint8 buf 0xd9;
    buf_add_uint8 buf len
  end
  else if len <= 0xffff then begin
    (* str 16 *)
    buf_add_uint8 buf 0xda;
    buf_add_uint16_be buf len
  end
  else begin
    (* str 32 *)
    buf_add_uint8 buf 0xdb;
    buf_add_uint32_be buf len
  end;
  Buffer.add_string buf s

let encode_bin buf b =
  let len = Bytes.length b in
  if len <= 0xff then begin
    (* bin 8 *)
    buf_add_uint8 buf 0xc4;
    buf_add_uint8 buf len
  end
  else if len <= 0xffff then begin
    (* bin 16 *)
    buf_add_uint8 buf 0xc5;
    buf_add_uint16_be buf len
  end
  else begin
    (* bin 32 *)
    buf_add_uint8 buf 0xc6;
    buf_add_uint32_be buf len
  end;
  Buffer.add_bytes buf b

let encode_array_header buf len =
  if len <= 15 then
    (* fixarray *)
    buf_add_uint8 buf (0x90 lor len)
  else if len <= 0xffff then begin
    (* array 16 *)
    buf_add_uint8 buf 0xdc;
    buf_add_uint16_be buf len
  end
  else begin
    (* array 32 *)
    buf_add_uint8 buf 0xdd;
    buf_add_uint32_be buf len
  end

let encode_map_header buf len =
  if len <= 15 then
    (* fixmap *)
    buf_add_uint8 buf (0x80 lor len)
  else if len <= 0xffff then begin
    (* map 16 *)
    buf_add_uint8 buf 0xde;
    buf_add_uint16_be buf len
  end
  else begin
    (* map 32 *)
    buf_add_uint8 buf 0xdf;
    buf_add_uint32_be buf len
  end

let rec encode_to_buf buf = function
  | Nil -> buf_add_uint8 buf 0xc0
  | Bool true -> buf_add_uint8 buf 0xc3
  | Bool false -> buf_add_uint8 buf 0xc2
  | Int n -> encode_int buf n
  | Str s -> encode_str buf s
  | Bin b -> encode_bin buf b
  | Array elts ->
      encode_array_header buf (List.length elts);
      List.iter (encode_to_buf buf) elts
  | Map pairs ->
      encode_map_header buf (List.length pairs);
      List.iter
        (fun (k, v) ->
          encode_to_buf buf k;
          encode_to_buf buf v)
        pairs

let encode v =
  let buf = Buffer.create 256 in
  encode_to_buf buf v;
  Buffer.to_bytes buf

(* -------------------------------------------------------------------------- *)
(* Decoding                                                                   *)
(* -------------------------------------------------------------------------- *)

let get_uint8 buf ofs =
  if ofs >= Bytes.length buf then failwith "msgpack: unexpected end of input";
  Char.code (Bytes.get buf ofs)

let get_uint16_be buf ofs =
  if ofs + 1 >= Bytes.length buf then
    failwith "msgpack: unexpected end of input";
  (Char.code (Bytes.get buf ofs) lsl 8) lor Char.code (Bytes.get buf (ofs + 1))

let get_uint32_be buf ofs =
  if ofs + 3 >= Bytes.length buf then
    failwith "msgpack: unexpected end of input";
  (Char.code (Bytes.get buf ofs) lsl 24)
  lor (Char.code (Bytes.get buf (ofs + 1)) lsl 16)
  lor (Char.code (Bytes.get buf (ofs + 2)) lsl 8)
  lor Char.code (Bytes.get buf (ofs + 3))

let get_int8 buf ofs =
  let v = get_uint8 buf ofs in
  if v >= 128 then v - 256 else v

let get_int16_be buf ofs =
  let v = get_uint16_be buf ofs in
  if v >= 32768 then v - 65536 else v

let get_int32_be buf ofs =
  let v = get_uint32_be buf ofs in
  (* On 64-bit OCaml, sign-extend from 32 bits *)
  if v > 0x7fffffff then v - (1 lsl 32) else v

let get_bytes buf ofs len =
  if ofs + len > Bytes.length buf then
    failwith "msgpack: unexpected end of input";
  Bytes.sub buf ofs len

let get_string buf ofs len =
  if ofs + len > Bytes.length buf then
    failwith "msgpack: unexpected end of input";
  Bytes.sub_string buf ofs len

let rec decode buf ofs =
  let tag = get_uint8 buf ofs in
  let ofs = ofs + 1 in
  if tag <= 0x7f then
    (* positive fixint *)
    (Int tag, ofs)
  else if tag >= 0xe0 then
    (* negative fixint *)
    (Int (tag - 256), ofs)
  else if tag >= 0xa0 && tag <= 0xbf then begin
    (* fixstr *)
    let len = tag land 0x1f in
    let s = get_string buf ofs len in
    (Str s, ofs + len)
  end
  else if tag >= 0x90 && tag <= 0x9f then begin
    (* fixarray *)
    let len = tag land 0x0f in
    decode_array buf ofs len
  end
  else if tag >= 0x80 && tag <= 0x8f then begin
    (* fixmap *)
    let len = tag land 0x0f in
    decode_map buf ofs len
  end
  else
    match tag with
    | 0xc0 -> (Nil, ofs)
    | 0xc2 -> (Bool false, ofs)
    | 0xc3 -> (Bool true, ofs)
    (* bin 8 *)
    | 0xc4 ->
        let len = get_uint8 buf ofs in
        let b = get_bytes buf (ofs + 1) len in
        (Bin b, ofs + 1 + len)
    (* bin 16 *)
    | 0xc5 ->
        let len = get_uint16_be buf ofs in
        let b = get_bytes buf (ofs + 2) len in
        (Bin b, ofs + 2 + len)
    (* bin 32 *)
    | 0xc6 ->
        let len = get_uint32_be buf ofs in
        let b = get_bytes buf (ofs + 4) len in
        (Bin b, ofs + 4 + len)
    (* uint 8 *)
    | 0xcc ->
        let v = get_uint8 buf ofs in
        (Int v, ofs + 1)
    (* uint 16 *)
    | 0xcd ->
        let v = get_uint16_be buf ofs in
        (Int v, ofs + 2)
    (* uint 32 *)
    | 0xce ->
        let v = get_uint32_be buf ofs in
        (Int v, ofs + 4)
    (* int 8 *)
    | 0xd0 ->
        let v = get_int8 buf ofs in
        (Int v, ofs + 1)
    (* int 16 *)
    | 0xd1 ->
        let v = get_int16_be buf ofs in
        (Int v, ofs + 2)
    (* int 32 *)
    | 0xd2 ->
        let v = get_int32_be buf ofs in
        (Int v, ofs + 4)
    (* str 8 *)
    | 0xd9 ->
        let len = get_uint8 buf ofs in
        let s = get_string buf (ofs + 1) len in
        (Str s, ofs + 1 + len)
    (* str 16 *)
    | 0xda ->
        let len = get_uint16_be buf ofs in
        let s = get_string buf (ofs + 2) len in
        (Str s, ofs + 2 + len)
    (* str 32 *)
    | 0xdb ->
        let len = get_uint32_be buf ofs in
        let s = get_string buf (ofs + 4) len in
        (Str s, ofs + 4 + len)
    (* array 16 *)
    | 0xdc ->
        let len = get_uint16_be buf ofs in
        decode_array buf (ofs + 2) len
    (* array 32 *)
    | 0xdd ->
        let len = get_uint32_be buf ofs in
        decode_array buf (ofs + 4) len
    (* map 16 *)
    | 0xde ->
        let len = get_uint16_be buf ofs in
        decode_map buf (ofs + 2) len
    (* map 32 *)
    | 0xdf ->
        let len = get_uint32_be buf ofs in
        decode_map buf (ofs + 4) len
    | _ ->
        failwith (Printf.sprintf "msgpack: unsupported format byte 0x%02x" tag)

and decode_array buf ofs len =
  let rec loop ofs acc remaining =
    if remaining = 0 then (Array (List.rev acc), ofs)
    else
      let v, ofs' = decode buf ofs in
      loop ofs' (v :: acc) (remaining - 1)
  in
  loop ofs [] len

and decode_map buf ofs len =
  let rec loop ofs acc remaining =
    if remaining = 0 then (Map (List.rev acc), ofs)
    else
      let k, ofs' = decode buf ofs in
      let v, ofs'' = decode buf ofs' in
      loop ofs'' ((k, v) :: acc) (remaining - 1)
  in
  loop ofs [] len

let decode_all buf =
  let total = Bytes.length buf in
  let rec loop ofs acc =
    if ofs >= total then List.rev acc
    else
      let v, ofs' = decode buf ofs in
      loop ofs' (v :: acc)
  in
  loop 0 []

(* -------------------------------------------------------------------------- *)
(* Convenience accessors                                                      *)
(* -------------------------------------------------------------------------- *)

let to_int = function Int n -> n | _ -> 0
let to_string = function Str s -> s | _ -> ""
let to_bool = function Bool b -> b | _ -> false
let to_list = function Array l -> l | _ -> []

let to_assoc = function
  | Map pairs ->
      List.filter_map
        (fun (k, v) -> match k with Str s -> Some (s, v) | _ -> None)
        pairs
  | _ -> []

let lookup key = function
  | Map pairs -> (
      match List.find_opt (fun (k, _) -> k = Str key) pairs with
      | Some (_, v) -> v
      | None -> Nil)
  | _ -> Nil

File: clients/zig/src/protocol.zig
/// protocol.zig -- Minimal msgpack encoder/decoder for kgd RPC.
///
/// Supports the subset of msgpack types needed by the protocol:
///   fixint, uint8/16/32, int8/16/32, nil, bool,
///   bin8/16/32, str8/16/32, fixstr,
///   fixarray, array16, fixmap, map16.
const std = @import("std");

// ---------- msgpack format byte constants ----------

const FMT_NIL: u8 = 0xc0;
const FMT_FALSE: u8 = 0xc2;
const FMT_TRUE: u8 = 0xc3;
const FMT_BIN8: u8 = 0xc4;
const FMT_BIN16: u8 = 0xc5;
const FMT_BIN32: u8 = 0xc6;
const FMT_UINT8: u8 = 0xcc;
const FMT_UINT16: u8 = 0xcd;
const FMT_UINT32: u8 = 0xce;
const FMT_INT8: u8 = 0xd0;
const FMT_INT16: u8 = 0xd1;
const FMT_INT32: u8 = 0xd2;
const FMT_FIXSTR_MASK: u8 = 0xa0;
const FMT_STR8: u8 = 0xd9;
const FMT_STR16: u8 = 0xda;
const FMT_STR32: u8 = 0xdb;
const FMT_FIXARRAY_MASK: u8 = 0x90;
const FMT_ARRAY16: u8 = 0xdc;
const FMT_FIXMAP_MASK: u8 = 0x80;
const FMT_MAP16: u8 = 0xde;

pub const EncodeError = error{OutOfMemory};

pub const DecodeError = error{
    UnexpectedEof,
    InvalidFormat,
    Overflow,
};

// ---------- Encoder ----------

/// Growable byte buffer that encodes msgpack values.
pub const Encoder = struct {
    buf: std.ArrayList(u8),

    pub fn init(allocator: std.mem.Allocator) Encoder {
        return .{ .buf = std.ArrayList(u8).init(allocator) };
    }

    pub fn deinit(self: *Encoder) void {
        self.buf.deinit();
    }

    /// Return the encoded bytes as a slice.  Caller does NOT own the memory
    /// (it belongs to the internal ArrayList).
    pub fn getWritten(self: *const Encoder) []const u8 {
        return self.buf.items;
    }

    /// Reset the buffer for reuse.
    pub fn reset(self: *Encoder) void {
        self.buf.clearRetainingCapacity();
    }

    // -- Primitive writers --

    pub fn writeNil(self: *Encoder) EncodeError!void {
        try self.buf.append(FMT_NIL);
    }

    pub fn writeBool(self: *Encoder, v: bool) EncodeError!void {
        try self.buf.append(if (v) FMT_TRUE else FMT_FALSE);
    }

    pub fn writeInt(self: *Encoder, v: i64) EncodeError!void {
        if (v >= 0) {
            try self.writeUint(@intCast(v));
        } else if (v >= -32) {
            // negative fixint: 111xxxxx
            try self.buf.append(@bitCast(@as(i8, @intCast(v))));
        } else if (v >= -128) {
            try self.buf.append(FMT_INT8);
            try self.buf.append(@bitCast(@as(i8, @intCast(v))));
        } else if (v >= -32768) {
            try self.buf.append(FMT_INT16);
            try self.appendBe16(@bitCast(@as(i16, @intCast(v))));
        } else {
            try self.buf.append(FMT_INT32);
            try self.appendBe32(@bitCast(@as(i32, @intCast(v))));
        }
    }

    pub fn writeUint(self: *Encoder, v: u64) EncodeError!void {
        if (v <= 0x7f) {
            // positive fixint
            try self.buf.append(@intCast(v));
        } else if (v <= 0xff) {
            try self.buf.append(FMT_UINT8);
            try self.buf.append(@intCast(v));
        } else if (v <= 0xffff) {
            try self.buf.append(FMT_UINT16);
            try self.appendBe16(@intCast(v));
        } else {
            try self.buf.append(FMT_UINT32);
            try self.appendBe32(@intCast(v));
        }
    }

    pub fn writeStr(self: *Encoder, s: []const u8) EncodeError!void {
        const len = s.len;
        if (len <= 31) {
            try self.buf.append(FMT_FIXSTR_MASK | @as(u8, @intCast(len)));
        } else if (len <= 0xff) {
            try self.buf.append(FMT_STR8);
            try self.buf.append(@intCast(len));
        } else if (len <= 0xffff) {
            try self.buf.append(FMT_STR16);
            try self.appendBe16(@intCast(len));
        } else {
            try self.buf.append(FMT_STR32);
            try self.appendBe32(@intCast(len));
        }
        try self.buf.appendSlice(s);
    }

    pub fn writeBin(self: *Encoder, data: []const u8) EncodeError!void {
        const len = data.len;
        if (len <= 0xff) {
            try self.buf.append(FMT_BIN8);
            try self.buf.append(@intCast(len));
        } else if (len <= 0xffff) {
            try self.buf.append(FMT_BIN16);
            try self.appendBe16(@intCast(len));
        } else {
            try self.buf.append(FMT_BIN32);
            try self.appendBe32(@intCast(len));
        }
        try self.buf.appendSlice(data);
    }

    pub fn writeArrayHeader(self: *Encoder, count: u32) EncodeError!void {
        if (count <= 15) {
            try self.buf.append(FMT_FIXARRAY_MASK | @as(u8, @intCast(count)));
        } else {
            try self.buf.append(FMT_ARRAY16);
            try self.appendBe16(@intCast(count));
        }
    }

    pub fn writeMapHeader(self: *Encoder, count: u32) EncodeError!void {
        if (count <= 15) {
            try self.buf.append(FMT_FIXMAP_MASK | @as(u8, @intCast(count)));
        } else {
            try self.buf.append(FMT_MAP16);
            try self.appendBe16(@intCast(count));
        }
    }

    // -- Helpers --

    fn appendBe16(self: *Encoder, v: u16) EncodeError!void {
        const bytes = std.mem.toBytes(std.mem.nativeToBig(u16, v));
        try self.buf.appendSlice(&bytes);
    }

    fn appendBe32(self: *Encoder, v: u32) EncodeError!void {
        const bytes = std.mem.toBytes(std.mem.nativeToBig(u32, v));
        try self.buf.appendSlice(&bytes);
    }
};

// ---------- Decoder ----------

/// A decoded msgpack value.
pub const Value = union(enum) {
    nil,
    boolean: bool,
    uint: u64,
    int: i64,
    str: []const u8,
    bin: []const u8,
    array: u32, // element count -- caller reads elements one by one
    map: u32, // entry count -- caller reads key/value pairs
};

/// Streaming msgpack decoder over a byte slice.
pub const Decoder = struct {
    data: []const u8,
    pos: usize = 0,

    pub fn init(data: []const u8) Decoder {
        return .{ .data = data, .pos = 0 };
    }

    /// Returns the number of bytes remaining.
    pub fn remaining(self: *const Decoder) usize {
        return self.data.len - self.pos;
    }

    /// Read the next value.
    pub fn read(self: *Decoder) DecodeError!Value {
        const b = try self.readByte();

        // positive fixint: 0xxxxxxx
        if (b <= 0x7f) return .{ .uint = b };

        // negative fixint: 111xxxxx
        if (b >= 0xe0) return .{ .int = @as(i8, @bitCast(b)) };

        // fixmap: 1000xxxx
        if (b & 0xf0 == 0x80) return .{ .map = b & 0x0f };

        // fixarray: 1001xxxx
        if (b & 0xf0 == 0x90) return .{ .array = b & 0x0f };

        // fixstr: 101xxxxx
        if (b & 0xe0 == 0xa0) {
            const len: usize = b & 0x1f;
            const s = try self.readSlice(len);
            return .{ .str = s };
        }

        return switch (b) {
            FMT_NIL => .nil,
            FMT_FALSE => .{ .boolean = false },
            FMT_TRUE => .{ .boolean = true },

            FMT_BIN8 => {
                const len: usize = try self.readByte();
                return .{ .bin = try self.readSlice(len) };
            },
            FMT_BIN16 => {
                const len: usize = try self.readU16();
                return .{ .bin = try self.readSlice(len) };
            },
            FMT_BIN32 => {
                const len: usize = try self.readU32();
                return .{ .bin = try self.readSlice(len) };
            },

            FMT_UINT8 => .{ .uint = try self.readByte() },
            FMT_UINT16 => .{ .uint = try self.readU16() },
            FMT_UINT32 => .{ .uint = try self.readU32() },

            FMT_INT8 => .{ .int = @as(i8, @bitCast(try self.readByte())) },
            FMT_INT16 => .{ .int = @as(i16, @bitCast(try self.readU16())) },
            FMT_INT32 => .{ .int = @as(i32, @bitCast(try self.readU32())) },

            FMT_STR8 => {
                const len: usize = try self.readByte();
                return .{ .str = try self.readSlice(len) };
            },
            FMT_STR16 => {
                const len: usize = try self.readU16();
                return .{ .str = try self.readSlice(len) };
            },
            FMT_STR32 => {
                const len: usize = try self.readU32();
                return .{ .str = try self.readSlice(len) };
            },

            FMT_ARRAY16 => .{ .array = try self.readU16() },
            0xdd => .{ .array = try self.readU32() }, // array32
            FMT_MAP16 => .{ .map = try self.readU16() },
            0xdf => .{ .map = try self.readU32() }, // map32

            else => DecodeError.InvalidFormat,
        };
    }

    /// Read a value and expect it to be an unsigned integer.
    pub fn readUint(self: *Decoder) DecodeError!u64 {
        const v = try self.read();
        return switch (v) {
            .uint => |u| u,
            .int => |i| if (i >= 0) @intCast(i) else DecodeError.InvalidFormat,
            else => DecodeError.InvalidFormat,
        };
    }

    /// Read a value and expect it to be a signed integer (accepts uint too).
    pub fn readInt(self: *Decoder) DecodeError!i64 {
        const v = try self.read();
        return switch (v) {
            .int => |i| i,
            .uint => |u| if (u <= @as(u64, @intCast(std.math.maxInt(i64)))) @intCast(u) else DecodeError.Overflow,
            else => DecodeError.InvalidFormat,
        };
    }

    /// Read a value and expect it to be a string.
    pub fn readStr(self: *Decoder) DecodeError![]const u8 {
        const v = try self.read();
        return switch (v) {
            .str => |s| s,
            else => DecodeError.InvalidFormat,
        };
    }

    /// Read a value and expect it to be a boolean.
    pub fn readBool(self: *Decoder) DecodeError!bool {
        const v = try self.read();
        return switch (v) {
            .boolean => |b| b,
            else => DecodeError.InvalidFormat,
        };
    }

    /// Read a value and expect it to be an array header. Returns element count.
    pub fn readArray(self: *Decoder) DecodeError!u32 {
        const v = try self.read();
        return switch (v) {
            .array => |n| n,
            else => DecodeError.InvalidFormat,
        };
    }

    /// Read a value and expect it to be a map header. Returns entry count.
    pub fn readMap(self: *Decoder) DecodeError!u32 {
        const v = try self.read();
        return switch (v) {
            .map => |n| n,
            else => DecodeError.InvalidFormat,
        };
    }

    /// Skip one complete value (including nested structures).
    pub fn skip(self: *Decoder) DecodeError!void {
        const v = try self.read();
        switch (v) {
            .array => |n| {
                var i: u32 = 0;
                while (i < n) : (i += 1) {
                    try self.skip();
                }
            },
            .map => |n| {
                var i: u32 = 0;
                while (i < n) : (i += 1) {
                    try self.skip(); // key
                    try self.skip(); // value
                }
            },
            else => {}, // scalars and str/bin already consumed
        }
    }

    /// Try to decode one complete top-level value starting at `pos`.
    /// Returns the byte length consumed, or null if the buffer is incomplete.
    pub fn tryMeasure(data: []const u8) ?usize {
        var d = Decoder.init(data);
        d.skip() catch return null;
        return d.pos;
    }

    // -- Internal byte reading --

    fn readByte(self: *Decoder) DecodeError!u8 {
        if (self.pos >= self.data.len) return DecodeError.UnexpectedEof;
        const b = self.data[self.pos];
        self.pos += 1;
        return b;
    }

    fn readSlice(self: *Decoder, len: usize) DecodeError![]const u8 {
        if (self.pos + len > self.data.len) return DecodeError.UnexpectedEof;
        const s = self.data[self.pos .. self.pos + len];
        self.pos += len;
        return s;
    }

    fn readU16(self: *Decoder) DecodeError!u16 {
        const s = try self.readSlice(2);
        return std.mem.bigToNative(u16, @bitCast(s[0..2].*));
    }

    fn readU32(self: *Decoder) DecodeError!u32 {
        const s = try self.readSlice(4);
        return std.mem.bigToNative(u32, @bitCast(s[0..4].*));
    }
};

// ---------- tests ----------

test "encode/decode nil" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeNil();
    var dec = Decoder.init(enc.getWritten());
    const v = try dec.read();
    try std.testing.expectEqual(Value.nil, v);
}

test "encode/decode bool" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeBool(true);
    try enc.writeBool(false);
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqual(true, (try dec.read()).boolean);
    try std.testing.expectEqual(false, (try dec.read()).boolean);
}

test "encode/decode positive fixint" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeUint(42);
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqual(@as(u64, 42), (try dec.read()).uint);
}

test "encode/decode negative fixint" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeInt(-5);
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqual(@as(i64, -5), (try dec.read()).int);
}

test "encode/decode uint8/16/32" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeUint(200);
    try enc.writeUint(30000);
    try enc.writeUint(100000);
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqual(@as(u64, 200), try dec.readUint());
    try std.testing.expectEqual(@as(u64, 30000), try dec.readUint());
    try std.testing.expectEqual(@as(u64, 100000), try dec.readUint());
}

test "encode/decode int8/16/32" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeInt(-100);
    try enc.writeInt(-1000);
    try enc.writeInt(-100000);
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqual(@as(i64, -100), try dec.readInt());
    try std.testing.expectEqual(@as(i64, -1000), try dec.readInt());
    try std.testing.expectEqual(@as(i64, -100000), try dec.readInt());
}

test "encode/decode fixstr" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeStr("hello");
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqualStrings("hello", try dec.readStr());
}

test "encode/decode str8" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    // Create a 40-byte string (> 31 so it uses str8).
    const long = "a" ** 40;
    try enc.writeStr(long);
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqualStrings(long, try dec.readStr());
}

test "encode/decode bin" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    const data = &[_]u8{ 0x89, 0x50, 0x4e, 0x47 };
    try enc.writeBin(data);
    var dec = Decoder.init(enc.getWritten());
    const v = try dec.read();
    try std.testing.expectEqualSlices(u8, data, v.bin);
}

test "encode/decode array" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeArrayHeader(3);
    try enc.writeUint(1);
    try enc.writeStr("two");
    try enc.writeBool(true);
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqual(@as(u32, 3), try dec.readArray());
    try std.testing.expectEqual(@as(u64, 1), try dec.readUint());
    try std.testing.expectEqualStrings("two", try dec.readStr());
    try std.testing.expectEqual(true, try dec.readBool());
}

test "encode/decode map" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeMapHeader(2);
    try enc.writeStr("key1");
    try enc.writeInt(42);
    try enc.writeStr("key2");
    try enc.writeStr("value");
    var dec = Decoder.init(enc.getWritten());
    try std.testing.expectEqual(@as(u32, 2), try dec.readMap());
    try std.testing.expectEqualStrings("key1", try dec.readStr());
    try std.testing.expectEqual(@as(i64, 42), try dec.readInt());
    try std.testing.expectEqualStrings("key2", try dec.readStr());
    try std.testing.expectEqualStrings("value", try dec.readStr());
}

test "skip nested structure" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    // [{"a": [1, 2]}, nil]
    try enc.writeArrayHeader(2);
    try enc.writeMapHeader(1);
    try enc.writeStr("a");
    try enc.writeArrayHeader(2);
    try enc.writeUint(1);
    try enc.writeUint(2);
    try enc.writeNil();
    // After skipping the entire structure, decoder should be at the end.
    var dec = Decoder.init(enc.getWritten());
    try dec.skip();
    try std.testing.expectEqual(@as(usize, 0), dec.remaining());
}

test "tryMeasure complete message" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeArrayHeader(2);
    try enc.writeUint(1);
    try enc.writeStr("ok");
    const written = enc.getWritten();
    try std.testing.expectEqual(written.len, Decoder.tryMeasure(written).?);
}

test "tryMeasure incomplete message" {
    var enc = Encoder.init(std.testing.allocator);
    defer enc.deinit();
    try enc.writeArrayHeader(3);
    try enc.writeUint(1);
    // Missing third element
    try std.testing.expectEqual(@as(?usize, null), Decoder.tryMeasure(enc.getWritten()));
}
